### PR TITLE
HEAL-73 & HEAL-74 - Doctor verification backend module

### PR DESCRIPTION
**Pull request by aishwaryaravi19**
Features:
HEAL-73 - Doctors should be able to raise credentials verification request - create and deploy smart contracts (Back End)
HEAL-74 - HealthCare Provider should be able to approve/reject verification request - create and deploy smart contracts (Back End)

Links:
https://codebrigadesjsu.atlassian.net/browse/HEAL-73
https://codebrigadesjsu.atlassian.net/browse/HEAL-74

### CODE REVIEW

File: contracts/HealthBlock.sol [hunk 1]
```diff
@@ -3,7 +3,7 @@ pragma solidity >=0.4.22 <0.9.0;
 
 //https://ethereum.stackexchange.com/questions/8615/child-contract-vs-struct?newreg=9940955131d740a1a85cef648b771ef3
 contract HealthBlock {
-    address private owner;
+    address private admin;
```
Comment (varunalla) [inline contracts/HealthBlock.sol]: owner is better, as it conforms to the blockchain nomenclature 
```suggestion
    address private owner;
```

File: contracts/HealthBlock.sol [hunk 2]
```diff
@@ -28,6 +28,67 @@ contract HealthBlock {
...
+        bool approved;
+    }
+    mapping(address => bool) private isDoctorVerified;
+    mapping(address => Request[]) private doctorRequests;
+
+   /*
+    * @dev Set contract deployer as owner
+    */
+    constructor() {
+        admin = msg.sender; // 'msg.sender' is sender of current call, contract deployer for a constructor
```
Comment (varunalla) [inline contracts/HealthBlock.sol]:
```suggestion
        owner = msg.sender; // 'msg.sender' is sender of current call, contract deployer for a constructor
```

File: contracts/HealthBlock.sol [hunk 3]
```diff
@@ -28,6 +28,67 @@ contract HealthBlock {
...
+    }
+
+    modifier onlyAdmin() {
+        require(msg.sender == admin, "Only admin can call this function");
```
Comment (varunalla) [inline contracts/HealthBlock.sol]:
```suggestion
        require(msg.sender == owner, "Only admin can call this function");
```

File: contracts/HealthBlock.sol [hunk 4]
```diff
@@ -28,6 +28,67 @@ contract HealthBlock {
...
+    event RequestApproved(address indexed doctor, address indexed provider, uint256 indexed requestId);
+    event RequestRejected(address indexed doctor, address indexed provider, uint256 indexed requestId);
+
+    struct Request {
+        address doctor;
+        string doctorName;
+        string credentialsHash;
+        bool approved;
+    }
+    mapping(address => bool) private isDoctorVerified;
```
Comment (varunalla) [inline contracts/HealthBlock.sol]: can we have this parameter of Doctor verified inside doctor details? with information regarding the entity who verified the doctor?

File: contracts/HealthBlock.sol [hunk 5]
```diff
@@ -28,6 +28,67 @@ contract HealthBlock {
...
+        require(!request.approved, "Request is already approved");
+        request.approved = true;
+        emit RequestApproved(request.doctor, admin, requestId);
+    }
+
+    function rejectRequest(uint256 requestId) public {
+        require(msg.sender == admin, "Only admin can call this function");
+        Request storage request = doctorRequests[admin][requestId];
+        require(!request.approved, "Request is already approved");
+        emit RequestRejected(request.doctor, admin, requestId);
```
Comment (varunalla) [inline contracts/HealthBlock.sol]: what does this method do? it does not update the status to rejected or remove the request?

File: contracts/HealthBlock.sol [hunk 6]
```diff
@@ -28,6 +28,67 @@ contract HealthBlock {
...
+    function getRequests() public view returns(Request[] memory) {
+        require(msg.sender == admin, "Only admin can call this function");
+        return doctorRequests[admin];
+    }
+
+    function approveRequest(uint256 requestId) public {
+        require(msg.sender == admin, "Only admin can call this function");
+        Request storage request = doctorRequests[admin][requestId];
+        require(!request.approved, "Request is already approved");
+        request.approved = true;
```
Comment (varunalla) [inline contracts/HealthBlock.sol:76]: can the request be status instead of true?

File: contracts/HealthBlock.sol [hunk 7]
```diff
@@ -28,6 +28,67 @@ contract HealthBlock {
...
+        require(msg.sender == admin, "Only admin can call this function");
+        _;
+    }
+
+    function verifyDoctor(address doctor) public onlyAdmin {
+        isDoctorVerified[doctor] = true;
+    }
+
+    function raiseRequest(string memory doctorName, string memory credentialsHash) public {
+        require(isDoctorVerified[msg.sender], "Only verified doctors can raise request");
```
Comment (varunalla) [inline contracts/HealthBlock.sol]: How will unverified doctor raise request to get verification done?

**Review by varunalla — CHANGES_REQUESTED** [review]

File: contracts/HealthBlock.sol (hunk 6 above)
Comment (aishwaryaravi19) [reply contracts/HealthBlock.sol:76]: this change has been made as part of integration ticket

**Review by varunalla — APPROVED** [review]
Approved!

**Pull request merged**